// @darkforest_eth/network/dist/TxExecutor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxExecutor = void 0;
const p_defer_1 = __importDefault(require("p-defer"));
const p_timeout_1 = __importDefault(require("p-timeout"));
const Network_1 = require("./Network");
const ThrottledConcurrentQueue_1 = require("./ThrottledConcurrentQueue");
class TxExecutor {
    constructor(ethConnection, gasSettingProvider, beforeTransaction, afterTransaction, queueConfiguration) {
        /**
         * Unless overridden, these are the default transaction options each blockchain transaction will
         * be sent with.
         */
        this.defaultTxOptions = {
            gasLimit: 2000000,
        };
        /**
         * Executes the given queued transaction. This is a field rather than a method declaration on
         * purpose for `this` purposes.
         */
        this.execute = async (txRequest) => {
            var _a;
            let time_called = undefined;
            let error = undefined;
            let time_submitted = undefined;
            let time_confirmed = undefined;
            let time_errored = undefined;
            let tx_hash = undefined;
            const time_exec_called = Date.now();
            try {
                await this.maybeUpdateNonce();
                if (this.beforeTransaction) {
                    await this.beforeTransaction(txRequest);
                }
                const requestWithDefaults = Object.assign(JSON.parse(JSON.stringify(this.defaultTxOptions)), txRequest.overrides);
                time_called = Date.now();
                const submitted = await (0, p_timeout_1.default)(txRequest.contract[txRequest.methodName](...txRequest.args, {
                    ...requestWithDefaults,
                    nonce: this.nonce,
                }), TxExecutor.TX_SUBMIT_TIMEOUT, `tx request ${txRequest.actionId} failed to submit: timed out}`);
                time_submitted = Date.now();
                tx_hash = submitted.hash;
                if (this.nonce !== undefined) {
                    this.nonce += 1;
                }
                this.lastTransactionTimestamp = time_submitted;
                txRequest.onTransactionResponse(submitted);

                this.eth.waitForTransaction(submitted.hash).then((confirmed) => {
                    time_confirmed = Date.now();
                    txRequest.onTransactionReceipt(confirmed);
            
                    // if (confirmed.status !== 1) {
                    //     time_errored = time_confirmed;
                    //     throw new Error('transaction reverted');
                    // }
                }).catch(txRequest.onReceiptError);

                // const confirmed = await this.ethConnection.waitForTransaction(submitted.hash);
                // if (confirmed.status !== 1) {
                //     time_errored = Date.now();
                //     throw new Error('transaction reverted');
                // }
                // else {
                //     time_confirmed = Date.now();
                //     txRequest.onTransactionReceipt(confirmed);
                // }
            }
            catch (e) {
                console.error(e);
                error = e;
                if (!time_submitted) {
                    time_errored = Date.now();
                    txRequest.onSubmissionError(error);
                }
                else {
                    // Ran out of retries, set nonce to undefined to refresh it
                    if (!time_errored) {
                        this.nonce = undefined;
                        time_errored = Date.now();
                    }
                    txRequest.onReceiptError(error);
                }
            }
            finally {
                (_a = this.diagnosticsUpdater) === null || _a === void 0 ? void 0 : _a.updateDiagnostics((d) => {
                    d.totalTransactions++;
                });
            }
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
            const logEvent = {
                tx_to: txRequest.contract.address,
                tx_type: txRequest.methodName,
                auto_gas_price_setting: txRequest.autoGasPriceSetting,
                time_exec_called,
                tx_hash,
            };
            if (time_called && time_submitted) {
                logEvent.wait_submit = time_submitted - time_called;
                if (time_confirmed) {
                    logEvent.wait_confirm = time_confirmed - time_called;
                }
            }
            if (error && time_errored) {
                logEvent.error = error.message || JSON.stringify(error);
                logEvent.wait_error = time_errored - time_exec_called;
                try {
                    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
                    if (error.body) {
                        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
                        logEvent.parsed_error = String.fromCharCode.apply(null, error.body || []);
                    }
                }
                catch (e) { }
            }
            logEvent.rpc_endpoint = this.ethConnection.getRpcEndpoint();
            logEvent.user_address = this.ethConnection.getAddress();
            this.afterTransaction && this.afterTransaction(txRequest, logEvent);
        };
        this.queue = new ThrottledConcurrentQueue_1.ThrottledConcurrentQueue(queueConfiguration !== null && queueConfiguration !== void 0 ? queueConfiguration : {
            invocationIntervalMs: 200,
            maxInvocationsPerIntervalMs: 3,
            maxConcurrency: 1,
        });
        this.lastTransactionTimestamp = Date.now();
        this.ethConnection = ethConnection;
        this.gasSettingProvider = gasSettingProvider;
        this.beforeTransaction = beforeTransaction;
        this.afterTransaction = afterTransaction;
    }
    /**
     * Schedules this transaction for execution.
     */
    queueTransaction(actionId, contract, methodName, args, overrides = {
        gasPrice: undefined,
        gasLimit: 2000000,
    }) {
        var _a;
        (_a = this.diagnosticsUpdater) === null || _a === void 0 ? void 0 : _a.updateDiagnostics((d) => {
            d.transactionsInQueue++;
        });
        const { promise: submittedPromise, reject: rejectTxResponse, resolve: txResponse, } = (0, p_defer_1.default)();
        const { promise: receiptPromise, reject: rejectTxReceipt, resolve: txReceipt, } = (0, p_defer_1.default)();
        const txRequest = {
            methodName,
            actionId,
            contract,
            args,
            overrides,
            onSubmissionError: rejectTxResponse,
            onReceiptError: rejectTxReceipt,
            onTransactionResponse: txResponse,
            onTransactionReceipt: txReceipt,
        };
        const autoGasPriceSetting = this.gasSettingProvider(txRequest);
        txRequest.autoGasPriceSetting = autoGasPriceSetting;
        if (overrides.gasPrice === undefined) {
            txRequest.overrides.gasPrice = (0, Network_1.gweiToWei)(this.ethConnection.getAutoGasPriceGwei(this.ethConnection.getAutoGasPrices(), autoGasPriceSetting));
        }
        this.queue.add(() => {
            var _a;
            (_a = this.diagnosticsUpdater) === null || _a === void 0 ? void 0 : _a.updateDiagnostics((d) => {
                d.transactionsInQueue--;
            });
            return this.execute(txRequest);
        });
        return {
            submitted: submittedPromise,
            confirmed: receiptPromise,
        };
    }
    /**
     * If the nonce is probably stale, reload it from the blockchain.
     */
    async maybeUpdateNonce() {
        if (this.nonce === undefined ||
            Date.now() - this.lastTransactionTimestamp > TxExecutor.NONCE_STALE_AFTER_MS) {
            const newNonce = await this.ethConnection.getNonce();
            if (newNonce !== undefined)
                this.nonce = newNonce;
        }
    }
    setDiagnosticUpdater(diagnosticUpdater) {
        this.diagnosticsUpdater = diagnosticUpdater;
    }
}
exports.TxExecutor = TxExecutor;
/**
 * A transaction is considered to have errored if haven't successfully submitted to mempool within
 * this amount of time.
 */
TxExecutor.TX_SUBMIT_TIMEOUT = 30000;
/**
 * We refresh the nonce if it hasn't been updated in this amount of time.
 */
TxExecutor.NONCE_STALE_AFTER_MS = 5000;
